### PR TITLE
AABB, OBB

### PR DESCRIPTION
**Pull request by super626**
Camera and Camera test 
plus
AABB, OBB

### CODE REVIEW

**Comment by super626** [comment]
Skip export AABB, OBB and Ray to lua.
some one should fix this.

File: cocos/3d/CCOBB.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    
+    /*
+     * Calculate the min and max project value of through the box's corners
+     */
+    void getInterval(const OBB& box, const Vec3& axis, float &min, float &max) const;
+    
+    /*
+     * Get the edege of x y z axis direction
+     */
+    Vec3 getEdgeDir(int index) const;
```
Comment (minggo) [inline cocos/3d/CCOBB.h]: It is better to use full word.
`DIR` may short of `directory` too.

File: cocos/3d/CCRay.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        
+		if (t > 0.f)
+		{
+			ptOnPlane = origin + t * dir;
+            
+			if (min.x < ptOnPlane.x && ptOnPlane.x < max.x && min.y < ptOnPlane.y && ptOnPlane.y < max.y)
+			{
+				return true;
+			}
+		}
```
Comment (minggo) [inline cocos/3d/CCRay.cpp]: Wrong indent.

File: cocos/3d/CCOBB.h (hunk 1 above)
Comment (super626) [inline cocos/3d/CCOBB.h]: Those functions are used internally. Do we need to change it?

Comment (minggo) [inline cocos/3d/CCOBB.h]: Yep, it is better to change it for others to read it.

File: cocos/3d/CCRay.cpp (hunk 2 above)
Comment (super626) [inline cocos/3d/CCRay.cpp]: Ok, thanks

File: cocos/3d/CCSprite3D.cpp [hunk 3]
```diff
@@ -399,4 +399,29 @@ const BlendFunc& Sprite3D::getBlendFunc() const
...
+    {
+        transform = getNodeToWorldTransform() * getSkin()->getRootBone()->getWorldMat();
+    }
+    else
+    {
+        transform = getNodeToWorldTransform();
+    }
+    
+    aabb.transform(transform);
+    return aabb;
```
Comment (minggo) [inline cocos/3d/CCSprite3D.cpp]: Can optimize it?
It is heavy to create aabb every time.

File: build/cocos2d_libs.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -8705,7 +8698,6 @@
 			isa = XCBuildConfiguration;
 			buildSettings = {
 				ALWAYS_SEARCH_USER_PATHS = YES;
-				ARCHS = "$(ARCHS_STANDARD)";
```
Comment (minggo) [inline build/cocos2d_libs.xcodeproj/project.pbxproj:8764]: Why modify it?

File: cocos/3d/CCMesh.h [hunk 5]
```diff
@@ -147,12 +153,21 @@ class CC_DLL Mesh : public Ref
     void buildBuffer();
     /**free buffer*/
     void cleanAndFreeBuffers();
+    
+    /*
+     * calculate AABB by origin vertices
+     * @param the vertices list
+     * @param stride the stride between two vertex's position data.
+     */
+    void calcuOriginAABB(const std::vector<float>& vertices, unsigned int stride);
```
Comment (minggo) [inline cocos/3d/CCMesh.h]: Is `calcu` well known as the short of `calculate`

File: cocos/3d/CCOBB.cpp [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+ THE SOFTWARE.
+ ****************************************************************************/
+
+#include "3d/CCOBB.h"
+
+NS_CC_BEGIN
+
+#define ROTATE(a,i,j,k,l) g=a.m[i + 4 * j]; h=a.m[k + 4 * l]; a.m[i + 4 * j]=(float)(g-s*(h+g*tau)); a.m[k + 4 * l]=(float)(h+s*(g-h*tau));
+
+static Mat4 _convarianceMatrix(const Vec3* aVertPos, int nVertCount)
```
Comment (minggo) [inline cocos/3d/CCOBB.cpp]: Function name should start with a verb.

Comment (minggo) [inline cocos/3d/CCOBB.cpp]: And don't use Hungary style, such as `nVerCount`.

File: cocos/3d/CCOBB.cpp [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+        _getElement(z, ip) = 0.0;
+    }
+
+    nrot = 0;
+
+    for(i=0; i<50; i++)
+    {
+        sm=0.0;
+        for(ip=0; ip<n; ip++) for(iq=ip+1;iq<n;iq++) sm+=fabs(a.m[ip + 4 * iq]);
+        if( sm == 0.0 )
```
Comment (minggo) [inline cocos/3d/CCOBB.cpp]: It is dangerous to compare it like is for `float`.

File: cocos/3d/CCOBB.cpp [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+        for(ip=0; ip<n; ip++)
+        {
+            for(iq=ip+1; iq<n; iq++)
+            {
+                g = 100.0*fabs(a.m[ip + iq * 4]);
+
+                float dmip = _getElement(d, ip);
+                float dmiq = _getElement(d, iq);
+                if( i>3 && fabs(dmip)+g==fabs(dmip) && fabs(dmiq)+g==fabs(dmiq) )
+                    a.m[ip + 4 * iq]=0.0;
```
Comment (minggo) [inline cocos/3d/CCOBB.cpp]: It is better to add `{}` for `if` because `else if` has `{}`

File: cocos/3d/CCOBB.cpp [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+        }
+    }
+
+    v.transpose();
+    *vout = v;
+    *dout = d;
+    return;
+}
+
+static Mat4 _GetOBBOrientation(const Vec3* aVertPos, int nVertCount)
```
Comment (minggo) [inline cocos/3d/CCOBB.cpp]: Don't use capital character for first word.

File: cocos/3d/CCOBB.cpp [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+            tmpLine.normalize();
+            break;
+        case 1:// edge with y axis
+            tmpLine = corners[7] - corners[6];
+            tmpLine.normalize();
+            break;
+        case 2:// edge with z axis
+            tmpLine = corners[1] - corners[6];
+            tmpLine.normalize();
+            break;
```
Comment (minggo) [inline cocos/3d/CCOBB.cpp:354]: How about add default?

File: cocos/3d/CCSprite3D.cpp (hunk 3 above)
Comment (lvlonggame) [inline cocos/3d/CCSprite3D.cpp]: cache the transform matrix?  that can optimize it a little.

File: build/cocos2d_libs.xcodeproj/project.pbxproj (hunk 4 above)
Comment (lvlonggame) [inline build/cocos2d_libs.xcodeproj/project.pbxproj:8764]: Ah, problem of merging code.

File: cocos/3d/CCOBB.cpp (hunk 7 above)
Comment (lvlonggame) [inline cocos/3d/CCOBB.cpp]: OK，i'll deal it.

File: cocos/3d/CCOBB.cpp (hunk 8 above)
Comment (lvlonggame) [inline cocos/3d/CCOBB.cpp]: OK!!

**Comment by zii** [comment]
What a strict master!

File: cocos/3d/CCOBB.h (hunk 1 above)
Comment (lvlonggame) [inline cocos/3d/CCOBB.h]: OK

File: cocos/3d/CCOBB.cpp (hunk 10 above)
Comment (lvlonggame) [inline cocos/3d/CCOBB.cpp:354]: OK

**Comment by minggo** [comment]
What's the status of this PR?
Is it ready for merging?

**Comment by lvlonggame** [comment]
yes! it's ready!

**Pull request merged**